ls1intum/Artemis
# Programming Exercise/Decrease number of REST calls on course-exercises for pending submissions + optimize logic for trigger-failed

### Checklist

### Motivation and Context



Fixes https://github.com/ls1intum/Artemis/issues/787.

### Description


**This PR fixes a couple of performance / logic issues for programming exercises in the courses area:**
- Instead of executing a REST call for every single course-exercies-row component to fetch the latest pending submission, the cache is initialized from the course-exercises page
- Because the course-exercises page only fetches the last graded result, we need to force clear the cache in the code-editor component. This means that we need to execute the REST call to retrieve a pending-submission every time the code-editor is opened (trade-off).
- The retrieved submission in the for-dashboard call was not correct for programming exercises as it would not retrieve a failed/pending submission (=no result), I also corrected this here
- Trigger failed now shows an error message for the student if it fails (see why in next point)
- The course-exercise-row now does not show the trigger-failed button for the latest submission anymore, but for the `latest graded submission`. This means that is likely however, that we can't re-trigger this build anymore, as the student might have made commits after the deadline. I didn't find an easy way to check this, so I return a 404 if the build result can't be matched from the CIS (only works if there have been no commits since the deadline).
- The updating-result component does not react anymore to ungraded submissions if showUngradedResults=false (inconsistency fix, otherwise it would show building becaues of an ungraded submission, but when the new result comes in it would show the old result again, because it is the last rated one)

**Known Bugs:**
- [ ] Course-exercise-details caches the complete result history in the participationWebsocketService. This way, ungraded submissions are set into the programmingSubmissionService. This issue causes the trigger-failed button to disappear until for-dashboard is loaded again.
The main problem here is that 'latest-pending-submission' does not consider re-triggered submissions to be pending as it does not know about the running build on the CIS.
- [ ] When triggering a failed build for an ungraded submission within the code-editor and then going to e.g. course-exercises and back to the code-editor, the building animation is lost. This is because the programmingSubmission cache is overriden with only graded submissions from the executed for-dashboard call on course-exercises (which this failed submission is not).

:information_source: The general problem here is that some components only care about graded submissions (e.g. result in course-exercise-row), while others about all submissions (e.g. code-editor). This can cause inconsistencies when submissions/results are passed into the programming-submission.service for caching.

### Steps for Testing


Given a programming exercise with a due date that runs out soon
1. Log in to Artemis
2. Navigate to Course Exercises
3. Create a couple of submissions in the code-editor before the due date passes, for the last one stop the build on Bamboo to create a failed submission
4. Create a couple of submissions in the code-editor after the due date has passed, for the last one stop the build on Bamboo to create a failed submission
5. Re-Trigger the failed submissions

Also when navigating to Course Exercises, no request to pending-submissions should be executed.


### Screenshots
